src/gamescope/wayland/manager.rs

```diff
@@ -0,0 +1,265 @@
+    ) {
+        // When receiving events from the wl_registry, we are only interested in the
+        // `global` event, which signals a new available global.
+        // When receiving this event, we just print its characteristics in this example.
+        if let wl_registry::Event::Global {
+            name,
+            interface,
+            version,
+        } = event
+        {
+            match &interface[..] {
+                "gamescope_control" => {
+                    log::debug!("Found gamescope control interface for Wayland!");
+                    let control = registry.bind::<GamescopeControl, _, _>(name, version, qh, ());
+                    state.control = Some(control);
+                }
+                "gamescope_input_method_manager" => {
+                    let input_method_manager =
+                        registry.bind::<GamescopeInputMethodManager, _, _>(name, version, qh, ());
+                    state.input_method_manager = Some(input_method_manager);
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+/// Handle events going to the [GamescopeControl] object.
+impl Dispatch<GamescopeControl, ()> for WaylandState {
+    fn event(
+        _state: &mut Self,
+        _control: &gamescope_control::GamescopeControl,
+        event: gamescope_control::Event,
+        _: &(),
+        _: &Connection,
+        _: &QueueHandle<WaylandState>,
+    ) {
+        match event {
+            gamescope_control::Event::FeatureSupport {
+                feature,
+                version,
+                flags,
+            } => {
+                log::debug!("Feature supported: {}, {}, {}", feature, version, flags);
+            }
+            gamescope_control::Event::ScreenshotTaken { path } => {
+                log::info!("Screenshot taken at path: {}", path);
+            }
+            _ => {}
+        }
+    }
+}
+
+/// Handle events going to the [GamescopeInputMethodManager] object.
+impl Dispatch<GamescopeInputMethodManager, ()> for WaylandState {
+    fn event(
+        _state: &mut Self,
+        _control: &gamescope_input_method_manager::GamescopeInputMethodManager,
+        _event: gamescope_input_method_manager::Event,
+        _: &(),
+        _: &Connection,
+        _: &QueueHandle<WaylandState>,
+    ) {
+    }
+}
+
+pub struct WaylandManager {
+    command_tx: Sender<WaylandMessage>,
+}
+
+impl WaylandManager {
+    pub async fn new() -> Result<Self, Box<dyn Error>> {
+        let (command_tx, command_rx) = tokio::sync::mpsc::channel::<WaylandMessage>(64);
+        let instance = Self { command_tx };
+
+        instance.run(command_rx).await?;
+
+        Ok(instance)
+    }
+
+    async fn run(&self, mut command_rx: Receiver<WaylandMessage>) -> Result<(), Box<dyn Error>> {
+        let runtime_dir =
+            env::var("XDG_RUNTIME_DIR").unwrap_or_else(|_| format!("/run/user/{}", Uid::current()));
+        let socket_path = format!("{}/gamescope-0", runtime_dir);
```

Comment:
In rare cases, it's possible that Gamescope will use `gamescope-1`, etc. instead of `gamescope-0`. We should instead discover this name instead of hard coding `gamescope-0`.